File: ruby/easy/sum_of_left_leaves.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Integer}
def sum_of_left_leaves(root)
    return calculate_sum(root, false)
end

def calculate_sum(root, is_left)
    return 0 unless root
    return root.val if (is_left && root.left.nil? && root.right.nil?)
    return calculate_sum(root.left, true) + calculate_sum(root.right, false)
end

File: javascript/easy/sumofLeftLeaves.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var sumOfLeftLeaves = function(root) {
    let queue = new Array(), sum = 0;
    if(root !== null) {
        queue.push(root);
    }
    while(queue.length > 0) {
        let current = queue.shift();
        if(current.left !== null) {
            if(current.left.left === null && current.left.right === null) {
                sum += current.left.val;
            }
            queue.push(current.left);
        }
        if(current.right !== null) {
            queue.push(current.right);
        }
    }
    return sum;
};
